src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -39,98 +40,234 @@ private ExpressionSyntax ParseInterpolatedStringToken()
+                        _ => throw ExceptionUtilities.UnexpectedValue(kind),
+                    },
+                    originalText[openQuoteRange],
+                    trailing: null);
             }
 
-            CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> getContent(ArrayBuilder<Lexer.Interpolation> interpolations)
+            CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> getContent(ReadOnlySpan<char> originalTextSpan)
             {
+                var content = PooledStringBuilder.GetInstance();
                 var builder = _pool.Allocate<InterpolatedStringContentSyntax>();
 
-                if (interpolations.Count == 0)
+                var indentationWhitespace = isMultiLineWithoutErrors ? getIndentationWhitespace(originalTextSpan) : default;
+
+                var currentContentStart = openQuoteRange.End;
+                for (var i = 0; i < interpolations.Count; i++)
                 {
-                    // In the special case when there are no interpolations, we just construct a format string
-                    // with no inserts. We must still use String.Format to get its handling of escapes such as {{,
-                    // so we still treat it as a composite format string.
-                    var text = originalText[new Range(openQuoteRange.End, closeQuoteRange.Start)];
-                    if (text.Length > 0)
-                    {
-                        builder.Add(SyntaxFactory.InterpolatedStringText(MakeInterpolatedStringTextToken(text, kind)));
-                    }
+                    var interpolation = interpolations[i];
+
+                    // Add a token for text preceding the interpolation
+                    builder.Add(makeContent(
+                        indentationWhitespace, content, first: i == 0, last: false,
+                        originalTextSpan[currentContentStart..interpolation.OpenBraceRange.Start])!);
+
+                    builder.Add(ParseInterpolation(this.Options, originalText, interpolation, kind));
+                    currentContentStart = interpolation.CloseBraceRange.End;
                 }
-                else
+
+                // Add a token for text following the last interpolation
+                builder.Add(makeContent(
+                    indentationWhitespace, content, first: interpolations.Count == 0, last: true,
+                    originalTextSpan[currentContentStart..closeQuoteRange.Start])!);
+
+                CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> result = builder;
+                _pool.Free(builder);
+                content.Free();
+                return result;
+            }
+
+            // Gets the indentation whitespace from the last line of a multi-line raw literal.
+            ReadOnlySpan<char> getIndentationWhitespace(ReadOnlySpan<char> originalTextSpan)
```

Comment:
Can we streamline the "multi-line with error" case, by just returning an empty span here? That will result in no dedentation without special logic elsewhere

Reply:
oh, that's nice!

Reply:
oh.  i do already do that :)  it's just in the caller.